stable/database/files/nuosm

```diff
@@ -517,6 +512,12 @@ log "$( nuocmd show archives --db-name $DB_NAME )"
 wotIdid="releasing first-in semaphore"
 [ -n "$first_in" ] && nuocmd set value --key $first_req --value '' --expected-value $HOSTNAME
 
+if [ -n "$restore_requested" ] && [ -n "$restore_successful" ]; then 
+    # clear the restore_request only if restore is successfull
+    trace "clearing restore request"
+    nuocmd set value --key $restore_req --value '' --expected-value "$restore_requested"
+fi
+
```

Comment:
I think this actually makes the timing issue less likely but does not eliminate it. In order to do that we would have to clear that entry in the KV store _after_ the initial SM is RUNNING, and then all other SMs can sync from it. But I don't know how you would do that if `nuodocker start sm` is "taking over".

Another approach would be to leave the entry untouched and add more intelligence to the check that the non-initial SMs perform when rejoining. The non-initial SMs should be able to join if there is already a running SM.

Of course, this would all be much simpler if it was implemented in `nuodocker` or in the Admin itself, but I think this process should be made more manual to eliminate this kind of complexity. Data recovery has to be performed with human intervention anyway, so why is there a bunch of complex logic to automate parts of it?

I think we should write down the steps that a user would follow to perform data recovery, and make sure we can support that. What I would prefer is something like the following,
1. Scale down SM StatefulSet to 1 replica.
2. Set some value in the KV store to pause the (one and only) SM.
3. Perform some explicit restore action on the SM filesystem, e.g. copy files, run `nuoarchive restore`, etc.
4. Unset the value in the KV store to unpause the (one and only) SM.
5. Wait for the SM to become RUNNING and then set the replica count to whatever was desired.

That would make it possible for us to diagnose issues with the data recovery process, which I think would be extremely difficult with all of this automatic logic that we now have.

Reply:
Yes, the initial change didn't solve the timing hazard completely as stated in the PR description. I've invested some more time to think about if this is possible using an asynchronous process that will wait for the database to become `RUNNING` before clearing the request in the KV store. Please check the latest changes.
If someone wants to have more control over the restore process, he/she can still scale up one SM at a time (similar to what the `TestKubernetesRestoreMultipleSMs/manualRestart` does).

This is how `in-place-restore` is triggered now:
```
helm install -n nuodb --generate-name stable/restore \
  --set restore.target=demo \
  --set restore.source="20210106T203316"
```

I agree that if we make the process more manual, we might remove the complexity that we currently have, however, this doesn't necessarily mean that the process will be less error-prone. In fact, I think that the most complex part of the `restore-in-place` is understanding and synchronizing the archives state in admin with the archives on disk which can be easily messed up.

I like the idea to also provide a "pause" option along with instructions in case someone wants to do the restore manually, however, I prefer to still support the automatic way. Happy to hear other comments on this.